frame/whitelist/src/lib.rs

```diff
@@ -0,0 +1,192 @@
+		type PreimageProvider: PreimageProvider<Self::Hash> + PreimageRecipient<Self::Hash>;
+
+		/// The weight information for this pallet.
+		type WeightInfo: weights::WeightInfo;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::generate_storage_info]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		CallWhitelisted { call_hash: T::Hash },
+		WhitelistedCallRemoved { call_hash: T::Hash },
+		WhitelistedCallDispatched { call_hash: T::Hash },
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		UnavailablePreImage,
+		UndecodableCall,
+		InvalidCallWeightWitness,
+		CallIsNotWhitelisted,
+		CallAlreadyWhitelisted,
+	}
+
+	#[pallet::storage]
+	pub type WhitelistedCall<T: Config> = StorageMap<_, Twox64Concat, T::Hash, (), OptionQuery>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(T::WeightInfo::whitelist_call())]
+		pub fn whitelist_call(origin: OriginFor<T>, call_hash: T::Hash) -> DispatchResult {
+			T::WhitelistOrigin::ensure_origin(origin)?;
+
+			ensure!(
+				!WhitelistedCall::<T>::contains_key(call_hash),
+				Error::<T>::CallAlreadyWhitelisted,
+			);
+
+			WhitelistedCall::<T>::insert(call_hash, ());
+			T::PreimageProvider::request_preimage(&call_hash);
+
+			Self::deposit_event(Event::<T>::CallWhitelisted { call_hash });
+
+			Ok(())
+		}
+
+		#[pallet::weight(T::WeightInfo::remove_whitelisted_call())]
+		pub fn remove_whitelisted_call(origin: OriginFor<T>, call_hash: T::Hash) -> DispatchResult {
+			T::WhitelistOrigin::ensure_origin(origin)?;
+
+			WhitelistedCall::<T>::take(call_hash).ok_or(Error::<T>::CallIsNotWhitelisted)?;
+
+			T::PreimageProvider::unrequest_preimage(&call_hash);
+
+			Self::deposit_event(Event::<T>::WhitelistedCallRemoved { call_hash });
+
+			Ok(())
+		}
+
+		#[pallet::weight(
+			T::WeightInfo::dispatch_whitelisted_call().saturating_add(*call_weight_witness)
+		)]
+		pub fn dispatch_whitelisted_call(
+			origin: OriginFor<T>,
+			call_hash: T::Hash,
+			call_weight_witness: Weight,
+		) -> DispatchResultWithPostInfo {
+			T::DispatchWhitelistedOrigin::ensure_origin(origin)?;
+
+			ensure!(
+				WhitelistedCall::<T>::contains_key(call_hash),
+				Error::<T>::CallIsNotWhitelisted,
+			);
+
+			let call = T::PreimageProvider::get_preimage(&call_hash)
+				.ok_or(Error::<T>::UnavailablePreImage)?;
+
+			let call = <T as Config>::Call::decode_with_depth_limit(
+				sp_api::MAX_EXTRINSIC_DEPTH,
+				&mut &call[..],
+			)
+			.map_err(|_| Error::<T>::UndecodableCall)?;
+
+			ensure!(
+				call.get_dispatch_info().weight <= call_weight_witness,
+				Error::<T>::InvalidCallWeightWitness
+			);
+
+			WhitelistedCall::<T>::remove(call_hash);
+
+			T::PreimageProvider::unrequest_preimage(&call_hash);
+
+			let result = call.dispatch(frame_system::Origin::<T>::Root.into());
+
+			Self::deposit_event(Event::<T>::WhitelistedCallDispatched { call_hash });
+
+			result
```

Comment:
We can't return result here since it could be error?

Instead we should return OK, and return the result in the event.

Reply:
ah indeed, should be fixed by https://github.com/paritytech/substrate/pull/10159/commits/4e80bae257c5fbc864c5fc8e40572aecf58a63a0